backend/models/viewpoint.js

```diff
@@ -16,167 +16,333 @@
-	if (data.screenshot && data.screenshot.base64) {
-		const croppedScreenshot = await utils.getCroppedScreenshotFromBase64(data.screenshot.base64, 79, 79);
-		data.screenshot.buffer = new Buffer.from(croppedScreenshot, "base64");
+		return viewpoint;
 	}
 
-	data = pick(data, ["name", "clippingPlanes", "viewpoint", "screenshot"]);
+	async create(sessionId, account, model, newView) {
+		if (!newView.name || !utils.isString(newView.name)) {
+			return Promise.reject({ resCode: responseCodes.INVALID_ARGUMENTS });
+		}
 
-	const newViewpoint = { _id:  utils.stringToUUID(nodeuuid()), ...data };
-	const coll = await db.getCollection(account, model + ".views");
-	await coll.insert(newViewpoint);
+		// ===============================
+		// DEPRECATED LEGACY SUPPORT START
+		// ===============================
+		if (newView.screenshot && newView.screenshot.base64) {
+			const croppedScreenshot = await utils.getCroppedScreenshotFromBase64(newView.screenshot.base64, 79, 79);
+			newView.thumbnail = new Buffer.from(croppedScreenshot, "base64");
+		}
 
-	const viewpoint = view.setViewpointThumbnailURL(account, model,  view.clean(newViewpoint));
+		if (newView.clippingPlanes && newView.viewpoint && !newView.viewpoint.clippingPlanes) {
+			newView.viewpoint.clippingPlanes = newView.clippingPlanes;
+		}
 
-	ChatEvent.viewpointsCreated(sessionId, account, model,viewpoint);
-	return viewpoint;
-};
+		delete newView.screenshot;
+		delete newView.clippingPlanes;
+		// =============================
+		// DEPRECATED LEGACY SUPPORT END
+		// =============================
+
+		newView._id = utils.stringToUUID(newView._id || nodeuuid());
+
+		if (newView.viewpoint) {
+			newView.viewpoint = await this.handleViewpoint(account, model, newView._id, newView.viewpoint);
+
+			if (newView.viewpoint.thumbnail) {
+				newView.thumbnail = newView.viewpoint.thumbnail;
+				delete newView.viewpoint.thumbnail;
+			}
+		}
 
-view.deleteViewpoint = function (account, model, idStr, sessionId) {
-	let id = idStr;
-	if (utils.isString(id)) {
-		id = utils.stringToUUID(id);
+		this.checkTypes(newView, this.fieldTypes);
+
+		const coll = await this.getCollection(account, model);
+		await coll.insert(newView);
+
+		newView = this.clean(account, model, newView);
+
+		if (sessionId) {
+			ChatEvent.viewpointsCreated(sessionId, account, model, newView);
+		}
+
+		return newView;
 	}
 
-	return db.getCollection(account, model + ".views").then((_dbCol) => {
-		return _dbCol.findOneAndDelete({ _id: id }).then((deleteResponse) => {
+	async deleteViewpoint(account, model, uid, sessionId) {
+		if (utils.isString(uid)) {
+			uid = utils.stringToUUID(uid);
+		}
+
+		const coll = await this.getCollection(account, model);
+		return coll.findOneAndDelete({ _id: uid }).then((deleteResponse) => {
 			if (!deleteResponse.value) {
-				return Promise.reject(responseCodes.VIEW_NOT_FOUND);
+				return Promise.reject(this.response("NOT_FOUND"));
```

Comment:
It can also return ISSUE_NOT_FOUND/RISK_NOT_FOUND - except you can't delete those things...

Reply:
Instead of relying on there's a response code called X_NOT_FOUND. maybe we should utilise inheritance and define a variable for these common error codes which gets overridden on the subclass?